Nikithauc/model as interfaces

nikithauc opened the pull request:
fixes #1013 
fixes #1521

Generated samples : https://github.com/microsoft/kiota-typescript/tree/nikithauc/sample-interfaces/packages/test/generatedCode/models/microsoft/graph

Graph v1 generation : https://github.com/microsoftgraph/msgraph-sdk-typescript/tree/nikithauc/sample-interfaces/src

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -88,37 +81,421 @@ public override void Refine(CodeNamespace generatedCode)
+
+    private static void ReplaceRequestConfigurationsQueryParamsWithInterfaces(CodeElement currentElement)
+    {
+        if (currentElement is CodeClass codeClass && (codeClass.IsOfKind(CodeClassKind.QueryParameters) || codeClass.IsOfKind(CodeClassKind.RequestConfiguration)))
```
```suggestion
        if (currentElement is CodeClass codeClass && codeClass.IsOfKind(CodeClassKind.QueryParameters, CodeClassKind.RequestConfiguration))
```

nikithauc replied:
Done

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -88,37 +81,421 @@ public override void Refine(CodeNamespace generatedCode)
+        {
+            currentClass.Name = currentClass.Name + ModelClassSuffix;
+        }
+        if (currentElement is CodeInterface modelInterface && modelInterface.IsOfKind(CodeInterfaceKind.Model))
+        {
+            if (modelInterface.Name.EndsWith("Interface"))
+            {
+                modelInterface.Name = modelInterface.Name.Split("Interface")[0];
+            }
+        }
```
```suggestion
        else if (currentElement is CodeInterface modelInterface && modelInterface.IsOfKind(CodeInterfaceKind.Model) && modelInterface.Name.EndsWith("Interface"))
        {
            modelInterface.Name = modelInterface.Name.Split("Interface")[0];
        }
```

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -88,37 +81,421 @@ public override void Refine(CodeNamespace generatedCode)
+        CrawlTree(currentElement, x => RenameModelInterfacesAndClasses(x));
+    }
+
+    private static void GenerateModelInterfaces(CodeElement currentElement, Func<CodeClass, string> interfaceNamingCallback)
```
instead of duplicating all of that, can't we reuse what we already have in the common refiner? even if it requires adding a few additional parameters

nikithauc replied:
There are multiple customizations specific to TypeScript design because of which it is convenient and readable to keep the logic separate. I started with that code in common language refiner and eventually maintained a separate one . There are specifics plugged in during the iteration of the usings, properties and methods to achieve the design 


Some examples are,
- While generating the interfaces they are temporarily named with `interface` suffix and finally this suffix is removed and add the "Impl" suffix to the model classes. 
- Usings are not removed in model class in TS 
- A parameter of model interface type is added in the model constructor.
- The return type of ReplaceType is different. 
- For request config and query parameters the classes are deleted from the namespace once the interfaces are generated.

baywet replied:
> While generating the interfaces they are temporarily named with interface suffix and finally this suffix is removed and add the "Impl" suffix to the model classes.

Why this temporary name?

> Usings are not removed in model class in TS

Could be driven by a boolean option

> A parameter of model interface type is added in the model constructor.

Could be driven by a boolean option

> The return type of ReplaceType is different.

Can you expand on that?

> For request config and query parameters the classes are deleted from the namespace once the interfaces are generated.

Could be driven by a boolean option

nikithauc replied:
The temporary name is because there is a condition that does not allow duplicate element names in the same namespace, that is, within the same DOM element. I appended the "Impl" after the generation so as to avoid missing the update to any references to the class using the class name.

baywet replied:
but why do you need a temporary name for the interface instead of 1. removing the class, 2. generating the interface, 3. renaming the class, 4. adding back both the interface and the class?

nikithauc replied:
> The return type of ReplaceType is different.

https://github.com/microsoft/kiota/blob/c9c6bd88bc89bfa67882b4ec36a1a4e4b0ea1e93/src/Kiota.Builder/Refiners/TypeScriptRefiner.cs#L607

This is different from the function in the `CommonRefiner ReplaceTypeByInterfaceType`.

nikithauc replied:
Created a separate issue for this https://github.com/microsoft/kiota/issues/1608.

I am currently focusing on https://github.com/microsoftgraph/msgraph-sdk-javascript/issues/824 and https://github.com/microsoft/kiota/issues/938 and there will updates in the design. 

It would be safer to unify the code after that.

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -88,37 +81,421 @@ public override void Refine(CodeNamespace generatedCode)
+            });
+        }
+    }
+    private static CodeInterface CreateModelInterface(CodeClass modelClass, Func<CodeClass, string> interfaceNamingCallback)
```
same comment about duplication

nikithauc replied:
Addressed above.

nikithauc commented:
Converting this to draft for now as I am investigating a bug where a some request executor method body parameters are not being replaced in the Graph V1 scope.

nikithauc commented:
https://github.com/microsoftgraph/msgraph-sdk-typescript/actions/runs/2368809346 - Graph V1 SDK successful runtime action

nikithauc commented:
> 

The sonar cloud comments have been resolved.

baywet commented on src/Kiota.Builder/CodeDOM/CodeProperty.cs:
```diff
@@ -56,4 +56,9 @@ public class CodeProperty : CodeTerminalWithKind<CodePropertyKind>, IDocumentedE
+
+    public object Clone()
+    {
+        return this.MemberwiseClone();
```
(also the class declaration needs to implement ICloneable)

nikithauc replied:
Resolved

baywet replied:
thanks, can you also call .clone on getter, setter and type to achieve deep cloning and avoid side effects please?

nikithauc commented on src/Kiota.Builder/CodeDOM/CodeInterface.cs:
```diff
@@ -1,13 +1,39 @@
 public class InterfaceDeclaration : ProprietableBlockDeclaration
 {
+
+    private readonly List<CodeType> inherits = new List<CodeType>();
```
An interface cannot have implements declaration and can extend multiple other interfaces in TS, Java.

interface A implements interface B // causes error
interface A extends B, C  // is valid 

Using implements structure would be ambiguous naming in that case.

baywet replied:
let me rephrase, all the other languages are using the implements structure to describe interfaces inheritance, we're adding that additional structure simply for TypeScript.

nikithauc replied:
Resolved.

nikithauc commented on src/Kiota.Builder/Writers/TypeScript/CodePropertyWriter.cs:
```diff
@@ -1,26 +1,44 @@
             default:
-                writer.WriteLine($"{conventions.GetAccessModifier(codeElement.Access)}{(codeElement.ReadOnly ? " readonly ": " ")}{codeElement.NamePrefix}{codeElement.Name.ToFirstCharacterLowerCase()}{(codeElement.Type.IsNullable ? "?" : string.Empty)}: {returnType}{(isFlagEnum ? "[]" : string.Empty)}{(codeElement.Type.IsNullable ? " | undefined" : string.Empty)};");
-            break;
+                if (codeElement.Parent is CodeInterface)
```
CodeElement.Parent can be a model class too.

baywet replied:
and you can have another default case with a when close. My argument is more about collapsing if cases in the switch cases to ease up readability.

nikithauc replied:
Changed the loop here

baywet commented:
thanks for making most of the changes. Could you also please:
- add a changelog entry
- put together a PR on the kiota-samples repo to update the typescript samples

nikithauc commented:
> thanks for making most of the changes. Could you also please:
> 
> * add a changelog entry
> * put together a PR on the kiota-samples repo to update the typescript samples

done

baywet commented:
Thanks for making the changes. I still believe we're missing an aspect: properties that are complex types or collections of complex types. if an event has a start date property of type date time time zone (also a model) when the JSON build object gets passed to it's constructor, if we have a value for that property, we should call the constructor for date time time zone and pass the value.

nikithauc commented:
> Thanks for making the changes. I still believe we're missing an aspect: properties that are complex types or collections of complex types. if an event has a start date property of type date time time zone (also a model) when the JSON build object gets passed to it's constructor, if we have a value for that property, we should call the constructor for date time time zone and pass the value.


Tried applying this suggestion. But this approach errors with the call stack exceeding error. I am afraid this introduces circular dependencies when initializing looped classes.

The serialization part already takes care of creating an instance. But doing that in the constructor errors out. 

https://github.com/microsoftgraph/msgraph-sdk-typescript/runs/7083070579?check_suite_focus=true#step:6:2454

I will open a new issue to test some scenarios based on this comment. It is a valid comment and the generation should securely map the code everytime.

baywet commented:
thanks for looking into that, setting the PR back to draft for now until we figure out a way out of circular dependencies

baywet commented:
Unless I missed something I don't believe properties that are complex types or collections of complex types are handled yet. If they are, can you also update the samples pr please?

nikithauc commented:
> Unless I missed something I don't believe properties that are complex types or collections of complex types are handled yet. If they are, can you also update the samples pr please?



> Unless I missed something I don't believe properties that are complex types or collections of complex types are handled yet. If they are, can you also update the samples pr please?

https://github.com/microsoft/kiota-samples/pull/776/files#diff-1b6ef19468094e83c07715397cc16851c9f69259095c5a2a17504e41ab30f68cR74 

https://github.com/microsoft/kiota-samples/pull/776/files#diff-6ffb85a3bf255b048e48eb5b7c799751e9d030081fd7773f1d793a0995fa9f9dR42

This is how a complex type is handled  at the serialization step.

baywet commented:
why does the model expect a class instance? the generic type constraint only specifies parsable.
https://github.com/microsoft/kiota-typescript/blob/91a90e3a71e0dfe93386ddb2c1e1ff540e9c9f1e/packages/abstractions/src/serialization/serializationWriter.ts#L91

The other affectation case you've shared is valid, I didn't think of that, thanks for sharing it.

What if we move that logic into the setter then instead? it should both satisfy the constructor and the affectation cases?

nikithauc commented:
> why does the model expect a class instance? the generic type constraint only specifies parsable. https://github.com/microsoft/kiota-typescript/blob/91a90e3a71e0dfe93386ddb2c1e1ff540e9c9f1e/packages/abstractions/src/serialization/serializationWriter.ts#L91
> 
> The other affectation case you've shared is valid, I didn't think of that, thanks for sharing it.
> 
> What if we move that logic into the setter then instead? it should both satisfy the constructor and the affectation cases?


1.  Example 

```
  writer.writeObjectValue<MessageRulePredicatesImpl>("exceptions", this.exceptions!);
```
`Error : Argument of type 'MessageRulePredicates' is not assignable to parameter of type 'MessageRulePredicatesImpl'` as a class instance may not necessarily be of a type or another implementation of the same interface. 


2. Regarding a setter, the code still needs to map each property value of the interface based object to the class instance property which causes the issue #1669 in the Graph V1 SDK. 


This PR is the first iteration. 

I expect more work to the models interfaces-classes aspects based on:
- #1669 
- #1662

nikithauc commented:
> > why does the model expect a class instance? the generic type constraint only specifies parsable. https://github.com/microsoft/kiota-typescript/blob/91a90e3a71e0dfe93386ddb2c1e1ff540e9c9f1e/packages/abstractions/src/serialization/serializationWriter.ts#L91
> > The other affectation case you've shared is valid, I didn't think of that, thanks for sharing it.
> > What if we move that logic into the setter then instead? it should both satisfy the constructor and the affectation cases?
> 
> 1. Example
> 
> ```
>   writer.writeObjectValue<MessageRulePredicatesImpl>("exceptions", this.exceptions!);
> ```
> 
> `Error : Argument of type 'MessageRulePredicates' is not assignable to parameter of type 'MessageRulePredicatesImpl'` as a class instance may not necessarily be of a type or another implementation of the same interface.
> 
> 2. Regarding a setter, the code still needs to map each property value of the interface based object to the class instance property which causes the issue [TypeScript - Investigate if nesting model class instances causes circular dependency #1669](https://github.com/microsoft/kiota/issues/1669) in the Graph V1 SDK.
> 
> This PR is the first iteration.
> 
> I expect more work to the models interfaces-classes aspects based on:
> 
> * [TypeScript - Investigate if nesting model class instances causes circular dependency #1669](https://github.com/microsoft/kiota/issues/1669)
> * [Create spec for TypeScript model usage using interfaces and classes.  #1662](https://github.com/microsoft/kiota/issues/1662)

Adding on the use of a setter, let's take the following example:

```
interface IUser
{
  drive :Drive;
}

 class UserImpl implements IUser{ 
    private _drive: Drive;
    public get drive() :Drive{
        return this._drive;
    }
    public set drive(value: Drive) {
    this._drive = value; 
  }
}

const userInstance = new UserImpl();

const userObject: IUser = {
  drive : new Drive()
}

console.log("Object : =>")
console.log(userObject);
console.log("Instance : =>")
console.log(userInstance);
```

Output: 

```
[LOG]: "Object : =>" 
[LOG]: {
  "drive": {}
} 
[LOG]: "Instance : =>" 
[LOG]: UserImpl: {
  "_drive": {}
} 
```

The property in the class instance and object differs. As the get/set does not persist in the JS instance, there is inconsistency between the model interface object and model instance. 

This is an issue when working with the model objects and instances. For example, when working with a task like Page Iteration, the code would have to access both response["values"] || response["_values"].

I remember @sebastienlevert raising this issue of the properties starting with "_" but the expectation is to not have them.

nikithauc commented:
Added model instantiation at the setter level

baywet commented:
Thanks for making the updates.
I'm not sure why we have this block?
https://github.com/microsoft/kiota-samples/pull/776/files/e21410bd705c5b182669b6e006693cd6821c6740..17dc949f99bfffc39fa4df1c99d36f8cfbe436ac#diff-06d4b04f846a062595d3f4c5a08d612f6e15d590a38991b97a1a0e4d151e40eaR74

And why is additional data nullable now? https://github.com/microsoft/kiota-samples/pull/776/files/e21410bd705c5b182669b6e006693cd6821c6740..17dc949f99bfffc39fa4df1c99d36f8cfbe436ac#diff-06d4b04f846a062595d3f4c5a08d612f6e15d590a38991b97a1a0e4d151e40eaR9

nikithauc commented:
> Thanks for making the updates. I'm not sure why we have this block? https://github.com/microsoft/kiota-samples/pull/776/files/e21410bd705c5b182669b6e006693cd6821c6740..17dc949f99bfffc39fa4df1c99d36f8cfbe436ac#diff-06d4b04f846a062595d3f4c5a08d612f6e15d590a38991b97a1a0e4d151e40eaR74

Which block are you referring to? 


> And why is additional data nullable now? https://github.com/microsoft/kiota-samples/pull/776/files/e21410bd705c5b182669b6e006693cd6821c6740..17dc949f99bfffc39fa4df1c99d36f8cfbe436ac#diff-06d4b04f846a062595d3f4c5a08d612f6e15d590a38991b97a1a0e4d151e40eaR9

I came across compilation issues with required additional data type . Example:

```
interface ITest{
    additionalData?:string
}

class test implements ITest{
    additionalData : string

    constructor(test:ITest){
        this.additionalData = test.additionalData
    }
}
```

```
Error: Type 'string | undefined' is not assignable to type 'string'.
  Type 'undefined' is not assignable to type 'string'
```


Also, when creating a request body using the model class, the additional data properties are not set. The additional data property has to be nullable especially when working with interfaces as the additional data becomes required at the time of populating a model object if it is not nullable

nikithauc commented:
> Why is the property defined as nullable in the model interface since it's not in the abstractions interface? https://github.com/microsoft/kiota-typescript/blob/04729081465baa983eec18d72f6853143c016f62/packages/abstractions/src/serialization/additionalDataHolder.ts#L7
> 
> Additional data should be set by the constructor.


1. If the `additionalData` is required in the model interface then every time a user creates a user object an error will be prompted asking for additionalData.
```
  const message: Message = {subject: "test"}; // This prompts an error if additionalData is required and not set here. 
``
If set as required,
```
 const message = {subject: "test", additionalData: {}}; // This is not ideal
```


2.
> this
> 
> ```ts
> if(this.value && this.value.length != 0){        const valueArrValue: AttachmentImpl[] = [];
>         this.value?.forEach(element => {
>             valueArrValue.push((element instanceof AttachmentImpl? element as AttachmentImpl:new AttachmentImpl(element)));
>         });
> ```


Added this since compilation fails without the proper casting.

baywet commented:
1. right, could we update the additional data holder interface in the abstractions to align things then please? 
2. if we changed this `writer.writeCollectionOfObjectValues<AttachmentImpl>("value", valueArrValue);` to that `writer.writeCollectionOfObjectValues<Attachment>("value", valueArrValue);` couldn't we do away with the block? (trying to avoid increasing the size)

nikithauc commented:
> 1. right, could we update the additional data holder interface in the abstractions to align things then please?
> 2. if we changed this `writer.writeCollectionOfObjectValues<AttachmentImpl>("value", valueArrValue);` to that `writer.writeCollectionOfObjectValues<Attachment>("value", valueArrValue);` couldn't we do away with the block? (trying to avoid increasing the size)


For point 2,

I run into compile time errors 

```
Types of property 'getFieldDeserializers' are incompatible.
    Type '(() => Record<string, (node: ParseNode) => void>) | undefined' is not assignable to type '() => Record<string, (node: ParseNode) => void>'.
```

But you bring up a valid point here. Created https://github.com/microsoft/kiota/issues/1730

baywet commented:
1. Thanks for making the change!
2. I believe this is because we have Partial<Parsable> which makes the method optional. We could get around that by updating the prototype of GetObjectValue, GetObjetCollectionValue in ParseNode I think.
3. I've noticed that the base types interfaces re-define the additionalData property, which is shouldn't since it's already defined in additional data holder, maybe add a filter in the refiner that creates the interfaces? https://github.com/microsoft/kiota-samples/blob/17dc949f99bfffc39fa4df1c99d36f8cfbe436ac/msgraph-mail/typescript/src/models/entity.ts#L5
4. The samples PR is conflicting for a lot of files. Maybe rebase interactively dropping your commits and re-generate?

nikithauc commented:
> 2\. I believe this is because we have Partial which makes the method optional. We could get around that by updating the prototype of GetObjectValue, GetObjetCollectionValue in ParseNode I think.

Not in this PR. I will this update the based on https://github.com/microsoft/kiota/issues/1732 and more design changes. 

Also note, this PR is already adding size optimization by replacing config param classes by interfaces.

> 3. I've noticed that the base types interfaces re-define the additionalData property, which is shouldn't since it's already defined in additional data holder, maybe add a filter in the refiner that creates the interfaces?

This is a separate issue. The changes in this PR are not redefining the additionalData property.

https://github.com/microsoft/kiota-samples/pull/776/files#diff-c9e4b2665b0a69f9959239648d9cc049770afd6b01f23154d8f296bf00c364e6L3

nikithauc commented:
> Are you able to provide numbers on the size reduction? I believe that because the classes end up being pulled by the fluent API surface for request bodies and by the discriminators for return types, and because the shape of the fluent API right now is hard to tree shake, it most likely makes only a little difference. Unless I'm missing something?

The reduction is in the unpacked disk size of the library because of using interfaces instead of classes. The lib size went from 64mb in preview 1 to 98 mb mainly with the request configurations. Converting the request configurations to interfaces removes the javascript transpiled classes and thereby reducing 
The bundle size optimization will require a drastic refactoring. At this point the aim is to optimize both the library size and the bundleable size of the fluent api.

baywet commented:
Alright, thanks for the additional information. Is this up to date or does it need to be refreshed at this point? https://github.com/microsoft/kiota-samples/pull/776

baywet reviewed (changes requested):
Thanks for updating this. There's a number of gaps we need to cover I've outlined in the samples PR we need to cover before we can merge this, otherwise a number of scenarios will break.
https://github.com/microsoft/kiota-samples/pull/776

baywet commented:
Hey @nikithauc
Thanks for continuing the work here.
I think there was a problem with your last commit. It's impacting far more files than it should (Ruby, Shell, MDs...). I'm not sure what commands you ran but I think and edit of this commit to only include TS files + a proper rebase is in order. Let me know if you need help with that. I want to make sure we don't loose any of your precious work while avoiding any regression in other "regions".

nikithauc commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -10,73 +10,68 @@
             AddParsableImplementsForModelClasses(generatedCode, "Parsable");
             ReplaceBinaryByNativeType(generatedCode, "ArrayBuffer", null, isNullable: true);
             cancellationToken.ThrowIfCancellationRequested();
             ReplaceReservedNames(generatedCode, new TypeScriptReservedNamesProvider(), x => $"{x}_escaped");
```
@baywet 
I see this error -> https://github.com/microsoft/kiota-typescript/actions/runs/4024090844/jobs/6915715774#step:14:18

In case the model name is a reserved word, such as `package` or `case`, then "sometimes" the `ReplaceReservedNames` function doesn't sync well with the `Interface conversion` process. As a result, `case_escaped`  and/or `package_escaped` remain as classes instead of converting to interface. This error is not consistent. There are times when I am able to refine and generate a complete set of model interfaces with `case_escaped`  and `package_escaped` as interfaces. 
 
 Do you have any suggestion to solve this issue?

baywet replied:
I'm not able to find `packages/test/generatedCode/models/package_escaped.ts` on the `nikithauc/model-interface-serialization` branch at the moment, did you remove it?
Could you provide the generated code for one of those failures?

nikithauc replied:
I did not delete this any artifact.
Pointing out to the inconsistency, if you see this run was successful https://github.com/microsoft/kiota-typescript/actions/runs/4033992745/jobs/6934865542. 
Let me see if I can generate this. There is no consistent way of reproducing this issue.

baywet replied:
so are you referring to a race condition? I'm struggling to understand what the issue is at this point.

nikithauc replied:
I believe it is a race condition. 

I was able to get an error sample reproduces. 
Check the following:

1. Case_escaped converts  fine to an interface: 
https://github.com/microsoft/kiota-typescript/blob/package-esc-missing/packages/test/generatedCode/models/security/case_escaped.ts

2. Package_escaped doesn't convert to an interface as expected. 
https://github.com/microsoft/kiota-typescript/blob/package-esc-missing/packages/test/generatedCode/models/package_escaped.ts

I believe the issue is  `ReplaceReservedNames` doesn't complete before `IntroducesInterfacesAndFunctions` synchronously. Sometimes both these classes fail to convert while sometimes they both convert fine.

baywet replied:
Refiners are single threaded and continuous to avoid race conditions at that stage. The only places where multiple threads are in use are when we're building the code dom in kiota builder (before the refiner's execution) and after when we write the files.
Ordering of the refiners can sometimes impact the result, it might be worth exploring that avenue.

nikithauc replied:
Resolved the issue. While replacing the reserved names, the `InnerChild` Elements of the Parent block were not updated with the new name. Added a fix for that in this  PR.

baywet commented on tests/Kiota.Builder.Tests/Refiners/TypeScriptLanguageRefinerTests.cs:
```diff
@@ -189,12 +177,8 @@ public async Task AddsUsingsForErrorTypesForRequestExecutor()
-            Kind = CodeClassKind.Model,
-            IsErrorDefinition = true,
-        }).First();
+        var errorClass = subNS.AddClass(TestHelper.CreateModelClass("rror4XX")).First();
```
```suggestion
        var errorClass = subNS.AddClass(TestHelper.CreateModelClass("Error4XX")).First();
```

baywet commented on tests/Kiota.Builder.Tests/Refiners/TypeScriptLanguageRefinerTests.cs:
```diff
@@ -267,24 +241,32 @@ public async Task AddsUsingsForDiscriminatorTypes()
         await ILanguageRefiner.Refine(new GenerationConfiguration { Language = GenerationLanguage.TypeScript }, root);
-        Assert.NotEqual("break", model.Name);
-        Assert.Contains("escaped", model.Name);
+        var interFaceModel = root.CodeInterfaces.First(x => x.Name == "Break_escaped");
+        Assert.NotEqual("break", interFaceModel.Name);
+        Assert.Contains("escaped", interFaceModel.Name);
```
```suggestion
        Assert.NotNull(root.CodeInterfaces.FirstOrDefault(x => "Break_escaped".Equals(x.Name, StringComparison.OrdinalIgnoreCase)));
```

baywet commented on src/Kiota.Builder/Writers/TypeScript/TypeScriptRelativeImportManager.cs:
```diff
@@ -38,7 +38,7 @@ public override (string, string, string) GetRelativeImportPathForUsing(CodeUsing
         else if (!isCodeUsingAModel)
         {
-            importPath += typeDef.Name.ToFirstCharacterLowerCase();
+            importPath += (codeUsing.Declaration?.TypeDefinition?.Name ?? codeUsing.Declaration?.Name).ToFirstCharacterLowerCase();
```
don't use null coalesce on strings

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeMethodWriter.cs:
```diff
@@ -188,7 +186,7 @@ private CodePropertyKind[] SetterAccessProperties
     private void WriteConstructorBody(CodeClass parentClass, CodeMethod currentMethod, LanguageWriter writer, bool inherits)
     {
-        if (inherits || parentClass.IsErrorDefinition)
+        if (!parentClass.IsOfKind(CodeClassKind.Model) && (inherits || parentClass.IsErrorDefinition))
```
can you provide more context on this change please?

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeMethodWriter.cs:
```diff
@@ -311,20 +299,32 @@ private void WriteRequestExecutorBody(CodeMethod codeElement, CodeClass parentCl
     {
         if (!codeElement.ReturnType.IsCollection) return fullTypeName;
-        var clone = (CodeTypeBase)codeElement.ReturnType.Clone();
+        var clone = codeElement.ReturnType.Clone() as CodeTypeBase;
+        if (clone == null)
+        {
+            throw new InvalidOperationException($"Something went wrong when cloning {codeElement.ReturnType.Name}");
+        }
```
clone always returns a value, not sure what this check adds?

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeMethodWriter.cs:
```diff
@@ -432,30 +442,10 @@ internal static void WriteMethodPrototypeInternal(CodeMethod code, LanguageWrite
             _ => string.Empty
         };
-        var shouldHaveTypeSuffix = !code.IsAccessor && !isConstructor;
+        var shouldHaveTypeSuffix = !code.IsAccessor && !isConstructor && returnType != ""; ;
```
```suggestion
        var shouldHaveTypeSuffix = !code.IsAccessor && !isConstructor && !string.IsNullOrEmpty(returnType);
```

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+
+    private void WriteSerializerFunction(CodeFunction codeElement, LanguageWriter writer)
+    {
+        var param = codeElement.OriginalLocalMethod.Parameters.FirstOrDefault(x => ((CodeType)x.Type).TypeDefinition is CodeInterface);
```
this will throw if it encounters a type that's a composed type. switch to type assertion instead of cast

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+
+            throw new InvalidOperationException("Interface parameter not found for code interface");
+
+        var codeInterface = (CodeInterface)codeType.TypeDefinition;
```
```suggestion
```

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+
+        if (inherits != null)
+        {
+            writer.WriteLine($"serialize{inherits?.TypeDefinition?.Name.ToFirstCharacterUpperCase()}(writer, {param.Name.ToFirstCharacterLowerCase()})");
```
```suggestion
            writer.WriteLine($"serialize{inherits.TypeDefinition!.Name.ToFirstCharacterUpperCase()}(writer, {param.Name.ToFirstCharacterLowerCase()})");
```

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+            WritePropertySerializer(codeInterface.Name.ToFirstCharacterLowerCase(), otherProp, writer, codeElement);
+        }
+
+        var additionalDataProperty = codeInterface.GetPropertyOfKind(CodePropertyKind.AdditionalData);
+        if (additionalDataProperty != null)
```
```suggestion
        if (codeInterface.GetPropertyOfKind(CodePropertyKind.AdditionalData) is CodeProperty additionalDataProperty)
```

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+        var codePropertyName = codeProperty.Name.ToFirstCharacterLowerCase();
+
+        var propertyTypeName = (codeProperty.Type as CodeType)?.TypeDefinition?.Name;
+        if (codeProperty.Parent == null)
+        {
+            throw new InvalidOperationException($"Code property must have parent {codePropertyName}");
+        }
```
should never be the case unless a unit test is poorly setup. Remove

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+
+        var serializationName = GetSerializationMethodName(codeProperty.Type);
+
+        if (serializationName == "writeObjectValue" || serializationName == "writeCollectionOfObjectValues")
```
move to a static readonly hashset with an ordinal ignore case comparer and use the contains method instead

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+    private string GetSerializationMethodName(CodeTypeBase propType)
+    {
+        var propertyType = localConventions?.TranslateType(propType);
+        if (propertyType != null && propType is CodeType currentType)
+        {
+            var result = GetSerializationMethodNameForCodeType(currentType, propertyType);
+            if (!String.IsNullOrWhiteSpace(result))
+            {
+                return result;
+            }
+        }
```
```suggestion
        if (!string.IsNullOrEmpty(propertyType) && propType is CodeType currentType && GetSerializationMethodNameForCodeType(currentType, propertyType) is string result && !String.IsNullOrWhiteSpace(result))
        {
            return result;
        }
```

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+
+    private static string? GetSerializationMethodNameForCodeType(CodeType propType, string propertyType)
+    {
+        var isCollection = propType.CollectionKind != CodeTypeBase.CodeTypeCollectionKind.None;
```
```suggestion
```

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+        var isCollection = propType.CollectionKind != CodeTypeBase.CodeTypeCollectionKind.None;
+        if (propType.TypeDefinition is CodeEnum currentEnum)
+            return $"writeEnumValue<{currentEnum.Name.ToFirstCharacterUpperCase()}>";
+        else if (isCollection)
```
```suggestion
        else if (propType.CollectionKind != CodeTypeBase.CodeTypeCollectionKind.None)
```

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+                    writer.WriteLine($"\"{keyName}\": n => {{ {param.Name.ToFirstCharacterLowerCase()}.{otherProp.Name.ToFirstCharacterLowerCase()} = n.{GetDeserializationMethodName(otherProp.Type, codeFunction)}; }},");
+                }
+            }
+            writer.DecreaseIndent();
+            writer.WriteLine("}");
```
```suggestion
            writer.CloseBlock();
```

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+            var properties = codeInterface?.Properties.Where(static x => x.Kind == CodePropertyKind.Custom && !x.ExistsInBaseType);
 
-        writer.WriteLine($"return new {returnType.ToFirstCharacterUpperCase()}();");
+            writer.WriteLine("return {");
```
```suggestion
            writer.StartBlock("return {");
```

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
 
-        writer.WriteLine($"return new {returnType.ToFirstCharacterUpperCase()}();");
+            writer.WriteLine("return {");
+            writer.IncreaseIndent();
```
```suggestion
```

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+        if (param?.Type is CodeType codeType && codeType.TypeDefinition is CodeInterface codeInterface)
+        {
+
+            var inherits = codeInterface?.StartBlock?.Implements.FirstOrDefault(x => x.TypeDefinition is CodeInterface);
```
```suggestion
```

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
-        writer.WriteLine($"return new {returnType.ToFirstCharacterUpperCase()}();");
+            writer.WriteLine("return {");
+            writer.IncreaseIndent();
+            if (inherits != null)
```
```suggestion
            if (codeInterface.StartBlock.Implements.FirstOrDefault(x => x.TypeDefinition is CodeInterface) is CodeInterface inherits)
```

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+
+            var inherits = codeInterface?.StartBlock?.Implements.FirstOrDefault(x => x.TypeDefinition is CodeInterface);
+
+            var properties = codeInterface?.Properties.Where(static x => x.Kind == CodePropertyKind.Custom && !x.ExistsInBaseType);
```
```suggestion
            var properties = codeInterface.Properties.Where(static x => x.Kind == CodePropertyKind.Custom && !x.ExistsInBaseType);
```

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+    {
+        var param = codeFunction.OriginalLocalMethod.Parameters.FirstOrDefault();
+
+        if (param?.Type is CodeType codeType && codeType.TypeDefinition is CodeInterface codeInterface)
```
```suggestion
        if (codeFunction.OriginalLocalMethod.Parameters.FirstOrDefault() is CodeParameter param && param.Type is CodeType codeType && codeType.TypeDefinition is CodeInterface codeInterface)
```

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+
+    private void WriteDeserializerFunction(CodeFunction codeFunction, LanguageWriter writer)
+    {
+        var param = codeFunction.OriginalLocalMethod.Parameters.FirstOrDefault();
+
```
```suggestion
```

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+                writer.WriteLine($"...deserializeInto{inherits?.TypeDefinition?.Name.ToFirstCharacterUpperCase()}({param.Name.ToFirstCharacterLowerCase()}),");
+            }
+
+            if (properties != null)
```
redundant if, remove

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+    {
+        var isCollection = propType.CollectionKind != CodeTypeBase.CodeTypeCollectionKind.None;
+        var propertyType = localConventions?.GetTypeString(propType, codeFunction, false);
+        if (propertyType != null && propType is CodeType currentType)
```
```suggestion
        if (!string.IsNullOrEmpty(propertyType) && propType is CodeType currentType)
```

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+        if (propertyType != null && propType is CodeType currentType)
+        {
+            if (currentType.TypeDefinition is CodeEnum currentEnum)
+                return $"getEnumValue{(currentEnum.Flags || isCollection ? "s" : string.Empty)}<{currentEnum.Name.ToFirstCharacterUpperCase()}>({propertyType.ToFirstCharacterUpperCase()})";
```
isn't the collection method getCollectionOfEnumValues? instead of just getEnumValues?

nikithauc replied:
Itis getEnumValues according to:

https://github.com/microsoft/kiota/blob/main/src/Kiota.Builder/Writers/TypeScript/CodeMethodWriter.cs
https://github.com/microsoft/kiota-typescript/blob/main/packages/abstractions/src/serialization/parseNode.ts

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+    {
+        var returnType = localConventions?.GetTypeString(targetClassType, currentElement, false);
+        var targetClassName = localConventions?.TranslateType(targetClassType);
+        var resultName = $"create{targetClassName.ToFirstCharacterUpperCase()}FromDiscriminatorValue";
```
this would fail if targetClassName is null, needs to be refactored

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +71,215 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+
+    private string? getSerializerAlias(CodeType propType, CodeFunction codeFunction, string propertySerializerName)
+    {
+        if (propType?.TypeDefinition?.Parent == null)
```
propType can't be null, refactor

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -293,4 +295,652 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+            };
+            targetNS.RemoveChildElement(codeClass);
+            var codeInterface = targetNS.AddInterface(insertValue).First();
+            var props = codeClass.Properties?.ToArray();
```
properties and usings can't be null, refactor

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -293,4 +295,652 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+
+        codeFunction.AddUsing(new CodeUsing
+        {
+            Name = modelInterface.Parent?.Name ?? string.Empty,
```
don't null coalesce strings

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -293,4 +295,652 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+            {
+                var targetNS = currentClass.GetImmediateParentOfType<CodeNamespace>();
+
+                var existing = targetNS.Classes.FirstOrDefault(x => x == currentClass);
+                if (existing == null)
```
```suggestion
                if (targetNS.Classes.FirstOrDefault(x => x == currentClass)== null)
```

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -293,4 +295,652 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+        /*
+         * Setting request body parameter type of request executor to model interface.
+         */
+        var requestBodyParam = codeMethod?.Parameters?.FirstOrDefault(x => x.Kind == CodeParameterKind.RequestBody);
```
code method can't be null, refactor

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -293,4 +295,652 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+            }
+
+            // Add the property of the model class to the model interface.
+            var newProperty = mProp.Clone() as CodeProperty;
+            if (newProperty != null)
```
```suggestion
            if (mProp.Clone() is CodeProperty newProperty)
```

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -293,4 +295,652 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+        originalType.TypeDefinition = propertyInterfaceType;
+        return (propertyInterfaceType, new CodeUsing
+        {
+            Name = propertyInterfaceType.Parent?.Name ?? string.Empty,
```
don't null coalesce strings

baywet reviewed (commented):
Thanks for pushing through this. A few trivial comments here and there.

nikithauc commented:
> Thanks for pushing through this. A few trivial comments here and there.

I have addressed the comments. A lot of the null checks were because the linter was prompting those errors and I added the null checks to bypass the errors. Used a null-forgiving operator.

gavinbarron commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -294,4 +296,647 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+
+        codeFunction.AddUsing(new CodeUsing
+        {
+            Name = modelInterface.Parent?.Name!,
```
Optional chaining and asserting a non-null value, this seems very wrong.

It strikes me that this is a case where we'd want to fail to generate, @baywet what's your take here?

baywet replied:
agreed, we should check whether the parent is not null before trying to add the using, and then the name will not be null

koros replied:
@baywet I see a few more instance where we're doing something like 
``` C#
serializer.AddUsing(new CodeUsing
{
    Name = parentSerializer.Parent?.Name!,
    Declaration = new CodeType
    {
        Name = parentSerializer.Name,
        TypeDefinition = parentSerializer
    }
});
```
should I convert everything to first check the parent is null then remove the optional chaining?

``` C#
if(parentSerializer.Parent != null && parentSerializer.Parent.Name != null)
{
	serializer.AddUsing(new CodeUsing
	{
		Name = parentSerializer.Parent?.Name!,
		Declaration = new CodeType
		{
			Name = parentSerializer.Name,
			TypeDefinition = parentSerializer
		}
	});
}
```

baywet replied:
yes please , however `if(parentSerializer.Parent != null)` should be enough because the name is never null with NRT

gavinbarron commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +72,196 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+        {
+            var resultName = $"create{targetClassName.ToFirstCharacterUpperCase()}FromDiscriminatorValue";
+            if (targetClassName.Equals(returnType, StringComparison.OrdinalIgnoreCase))
+                return $"{resultName}";
```
there's no need to interpolate here

baywet reviewed (changes requested):
Thanks for making the changes, here are some additional comments

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -1,31 +1,57 @@
-
+    private TypeScriptConventionService? localConventions;
+    private readonly CodeUsingWriter _codeUsingWriter;
+    private static HashSet<string> customSerializatioWriters = new HashSet<string> { "writeObjectValue", "writeCollectionOfObjectValues" };
```
```suggestion
    private static readonly HashSet<string> customSerializatioWriters = new HashSet<string>(StringComparison.OrdinalIgnoreCase) { "writeObjectValue", "writeCollectionOfObjectValues" };
```

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -294,4 +294,649 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+         */
+        if (codeMethod.Parameters.FirstOrDefault(static x => x.Kind == CodeParameterKind.RequestBody) is CodeParameter requestBodyParam)
+        {
+            if (requestBodyParam?.Type is CodeType requestBodyType && requestBodyType.TypeDefinition is CodeClass requestBodyClass)
```
merge if with enclosing one

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -294,4 +294,649 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+        {
+            if (requestBodyParam?.Type is CodeType requestBodyType && requestBodyType.TypeDefinition is CodeClass requestBodyClass)
+            {
+                SetTypeAsModelInterface(CreateModelInterface(requestBodyClass, interfaceNamingCallback), (CodeType)requestBodyParam.Type, requestBuilderClass);
```
```suggestion
                SetTypeAsModelInterface(CreateModelInterface(requestBodyClass, interfaceNamingCallback), requestBodyType, requestBuilderClass);
```

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -294,4 +294,649 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+                        AddSerializationUsingToRequestBuilder(requestBodyClass, parentClass);
+                    }
+
+                    if (parentClass.Name != requestBodyClass.Name)
```
use equals with a string comparison instead

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -294,4 +294,649 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+            {
+                SetTypeAsModelInterface(CreateModelInterface(requestBodyClass, interfaceNamingCallback), (CodeType)requestBodyParam.Type, requestBuilderClass);
+
+                if (codeMethod?.Kind == CodeMethodKind.RequestGenerator)
```
```suggestion
                if (codeMethod.Kind == CodeMethodKind.RequestGenerator)
```

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -294,4 +294,649 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+                }
+            }
+        }
+        if (codeMethod?.ErrorMappings != null && codeMethod.ErrorMappings.Any())
```
```suggestion
        if (codeMethod.ErrorMappings.Any())
```

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -294,4 +294,649 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+            Kind = CodeInterfaceKind.Model,
+        };
+
+        var modelInterface = (shouldInsertUnderParentClass ?
+                       modelParentClass?.AddInnerInterface(insertValue).First() :
+                       namespaceOfModel.AddInterface(insertValue).First()) ?? throw new InvalidOperationException("Model interface can't be null");
```
```suggestion
        var modelInterface = (modelClass.Parent is CodeClass modelParentClass ?
                       modelParentClass.AddInnerInterface(insertValue).First() :
                       namespaceOfModel.AddInterface(insertValue).First());
```

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -294,4 +294,649 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+
+    private static void SetUsingInModelInterface(CodeInterface modelInterface, (CodeInterface?, CodeUsing?) propertyTypeAndUsing)
+    {
+        if (propertyTypeAndUsing.Item1 is not null && propertyTypeAndUsing.Item2 is not null && modelInterface.Name != propertyTypeAndUsing.Item1.Name)
```
string comparison instead

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -294,4 +294,649 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+
+    private static void SetUsingsOfPropertyInSerializationFunctions(string propertySerializerFunctionName, CodeFunction codeFunction, CodeClass property, Func<CodeClass, string> interfaceNamingCallback)
+    {
+        if (!propertySerializerFunctionName.Equals(codeFunction.Name))
```
```suggestion
        if (!propertySerializerFunctionName.Equals(codeFunction.Name, StringComparison.OrdinalIgnoreCase))
```

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -294,4 +294,649 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+    {
+        if (!propertySerializerFunctionName.Equals(codeFunction.Name))
+        {
+            var serializationFunction = GetSerializationFunctionsForNamespace(property).Item1 ?? throw new InvalidOperationException($"Serialization function for property {property.Name} not found");
+            if (serializationFunction.Parent is not null)
```
```suggestion
            if (GetSerializationFunctionsForNamespace(property).Item1 is CodeFunction serializationFunction && serializationFunction.Parent is not null)
```

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -294,4 +294,649 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+
+    protected static void AddStaticMethodsUsingsToDeserializerFunctions(CodeElement currentElement, Func<CodeType, string> functionNameCallback)
+    {
+        if (currentElement is CodeFunction codeFunction && codeFunction.OriginalLocalMethod is CodeMethod currentMethod && currentMethod.Kind == CodeMethodKind.Deserializer)
+        {
+            var interfaceParameter = currentMethod.Parameters.FirstOrDefault(x => x.Type is CodeType codeType && codeType.TypeDefinition is CodeInterface);
+
+            if (interfaceParameter?.Type is CodeType codeType && codeType.TypeDefinition is CodeInterface ci && ci.GetChildElements(true).OfType<CodeProperty>() is IEnumerable<CodeProperty> props)
+            {
+                foreach (var property in props)
+                {
+                    AddPropertyFactoryUsingToDeserializer(codeFunction, property, functionNameCallback);
+                }
+            }
+        }
```
```suggestion
        if (currentElement is CodeFunction codeFunction && codeFunction.OriginalLocalMethod is CodeMethod currentMethod && currentMethod.Kind == CodeMethodKind.Deserializer && currentMethod.Parameters.FirstOrDefault(x => x.Type is CodeType codeType && codeType.TypeDefinition is CodeInterface) is CodeParameter interfaceParameter && interfaceParameter?.Type is CodeType codeType && codeType.TypeDefinition is CodeInterface ci)
        {
            foreach (var property in ci.Properties)
            {
                AddPropertyFactoryUsingToDeserializer(codeFunction, property, functionNameCallback);
            }
        }
```

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -294,4 +294,649 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+    private static void AddDeserializerUsingToDisriminatorFactory(CodeElement codeElement)
+    {
+        if (codeElement is CodeFunction parsableFactoryFunction && parsableFactoryFunction.OriginalLocalMethod.Kind == CodeMethodKind.Factory
+            && parsableFactoryFunction.OriginalLocalMethod?.ReturnType is CodeType codeType && codeType?.TypeDefinition is CodeClass modelReturnClass)
```
```suggestion
            && parsableFactoryFunction.OriginalLocalMethod?.ReturnType is CodeType codeType && codeType.TypeDefinition is CodeClass modelReturnClass)
```

baywet reviewed (changes requested):
Thanks for making all those changes. Getting there.

koros commented:
@baywet Please have another look, I think I've addressed most of the requested changes

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +72,189 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+        var isCollectionOfEnum = IsCodePropertyCollectionOfEnum(codeProperty);
+        var spreadOperator = isCollectionOfEnum ? "..." : string.Empty;
+        var codePropertyName = codeProperty.Name.ToFirstCharacterLowerCase();
+
+        var propertyTypeName = (codeProperty.Type as CodeType)?.TypeDefinition?.Name;
+
```
```suggestion
```

baywet commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +72,196 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+
+        if (customSerializatioWriters.Contains(serializationName))
+        {
+            var serializeName = getSerializerAlias((codeProperty.Type as CodeType)!, codeFunction, $"serialize{propertyTypeName}");
```
```suggestion
            var serializeName = getSerializerAlias(propType, codeFunction, $"serialize{propType.Name}");
```

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -294,4 +294,635 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+
+    private static void SetUsingInModelInterface(CodeInterface modelInterface, (CodeInterface?, CodeUsing?) propertyTypeAndUsing)
+    {
+        if (propertyTypeAndUsing.Item1 is not null && propertyTypeAndUsing.Item2 is not null && !modelInterface.Name.Equals(propertyTypeAndUsing.Item1.Name))
```
```suggestion
        if (propertyTypeAndUsing.Item1 is not null && propertyTypeAndUsing.Item2 is not null && !modelInterface.Name.Equals(propertyTypeAndUsing.Item1.Name, StringComparison.OrdinalIgnoreCase))
```

baywet commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -294,4 +294,635 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+
+    private static void SetUsingsOfPropertyInSerializationFunctions(string propertySerializerFunctionName, CodeFunction codeFunction, CodeClass property, Func<CodeClass, string> interfaceNamingCallback)
+    {
+        if (!propertySerializerFunctionName.EqualsIgnoreCase(codeFunction.Name))
```
```suggestion
        if (!propertySerializerFunctionName.EqualsIgnoreCase(codeFunction.Name, StringComparison.OrdinalIgnoreCase))
```

baywet commented on tests/Kiota.Builder.Tests/Refiners/TypeScriptLanguageRefinerTests.cs:
```diff
@@ -51,34 +53,27 @@ public async Task AddsQueryParameterMapperMethod()
         {
-            Name = "Deserialize",
-            Kind = CodeMethodKind.Deserializer,
+            Name = "factory",
+            Kind = CodeMethodKind.Factory,
```
I don't think the kind and name changes are required here anymore, I think that change came from a previous state of the code that's no longer require (I can't find any trace in the refiner of a method changing the kind from deserializer to factory)

baywet reviewed (dismissed):
Thank you so much for pulling through all the feedback!!!
LGTM. After integration tests are done, I'm happy to get this one merged.
Setting automerge now so we don't forget about squashing later on.

baywet commented on src/Kiota.Builder/Refiners/CommonLanguageRefiner.cs:
```diff
@@ -405,6 +405,18 @@ protected static void ReplaceReservedNames(CodeElement current, IReservedNamesPr
         CrawlTree(current, x => ReplaceReservedNames(x, provider, replacement, codeElementExceptions, shouldReplaceCallback));
     }
+
+    private static void UpdateReservedNameReplacementInParent(CodeElement codeElement, string newDeclarationName)
```
this method is not used anywhere, remove

baywet commented:
@andrueastman @rkodev would you mind giving it a look just for sanity please?

andrueastman commented on src/Kiota.Builder/Writers/TypeScript/TypeScriptConventionService.cs:
```diff
@@ -119,9 +120,19 @@ public override string TranslateType(CodeType type)
+
+    private static string GetCodeTypeName(CodeType codeType)
+    {
+        if (codeType.TypeDefinition is CodeFunction)
+        {
+            return !string.IsNullOrEmpty(codeType.TypeDefinition?.Name) ? codeType.TypeDefinition.Name : string.Empty;
+        }
```
```suggestion
        if (codeType.TypeDefinition is CodeFunction codeFunction)
        {
            return !string.IsNullOrEmpty(codeFunction.Name) ? codeType.TypeDefinition.Name : string.Empty;
        }
```

andrueastman commented on src/Kiota.Builder/Writers/TypeScript/CodeMethodWriter.cs:
```diff
@@ -362,30 +356,34 @@ private void WriteRequestGeneratorBody(CodeMethod codeElement, RequestParams req
-        if (additionalDataProperty != null)
-            writer.WriteLine($"writer.writeAdditionalData(this.{additionalDataProperty.Name.ToFirstCharacterLowerCase()});");
     }
+    private static string GetPropertyCall(CodeProperty property, string defaultValue) => property == null ? defaultValue : $"this.{property.Name}";
```
It looks like `property` will never be null here. All usages pass a non null property and the parameter should probably have the nullable annotation as `CodeProperty?` if this is possible.

andrueastman commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -1,31 +1,57 @@
+
         writer.IncreaseIndent();
+
+        localConventions = new TypeScriptConventionService(writer);
```
I'm not sure if we should be re-initializing the conventions every time `WriteCodeElement` is called. This could potentially be unnecessarily expensive.

The base type also provides the `conventions` property that is initialized in the constructor. Any reason we can't replace `localConventions` with `conventions` from the base? 
(If this is because of the writer, I believe the `GetTypeString` function allows you to pass the writer as an optional param rather than creating a new Conventions instance every time)

andrueastman commented on src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs:
```diff
@@ -46,17 +72,186 @@ private void WriteFactoryMethodBody(CodeFunction codeElement, string returnType,
+            if (propType.TypeDefinition == null)
+                return $"writeCollectionOfPrimitiveValues<{propertyType.ToFirstCharacterLowerCase()}>";
+            else
+                return $"writeCollectionOfObjectValues";
```
```suggestion
                return "writeCollectionOfObjectValues";
```

andrueastman commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -307,4 +307,631 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+            var usings = codeClass.Usings.ToArray();
+            codeInterface.AddUsing(usings);
+        }
+        CrawlTree(currentElement, x => ReplaceRequestConfigurationsQueryParamsWithInterfaces(x));
```
```suggestion
        CrawlTree(currentElement, static x => ReplaceRequestConfigurationsQueryParamsWithInterfaces(x));
```

andrueastman commented on src/Kiota.Builder/Refiners/TypeScriptRefiner.cs:
```diff
@@ -307,4 +307,631 @@ private static void CorrectMethodType(CodeMethod currentMethod)
+        {
+            CreateSerializationFunctions(codeClass);
+        }
+        CrawlTree(codeElement, x => CreateSeparateSerializers(x));
```
```suggestion
        CrawlTree(codeElement, static x => CreateSeparateSerializers(x));
```

andrueastman reviewed (approved):
Thanks for all this. Looks good to me.
